designs/2021-suppression-support/README.md

```diff
@@ -0,0 +1,511 @@
+[
+  {
+    "filePath":"C:\\Users\\yiweiding\\Documents\\projects\\test\\1.js",
+    "messages":[
+      {
+        "ruleId":"no-undef",
+        "severity":2,
+        "message":"'a' is not defined.",
+        "line":1,
+        "column":1,
+        "nodeType":"Identifier",
+        "suppressions":[
+          {
+            "kind":"inSource",
+            "justification":"Sample justification message."
+          }
+        ],
+        "messageId":"undef",
+        "endLine":1,
+        "endColumn":2
+      }
+    ],
+    "errorCount":0,
+    "warningCount":0,
+    "fixableErrorCount":0,
+    "fixableWarningCount":0,
+    "usedDeprecatedRules":[]
+  }
+]
+```
+
+### Add new CLI option `--output-suppression`
+
+We propose to add a new CLI option `--output-suppression` for the new feature. Without the option, ESLint will work as it is currently, and `suppressions` property should not exist in the output. With this option, ESLint is expected to export all violations, including suppressed ones. `suppressions` should be _empty_ for non-suppressed violations, while `suppressions` should be a list for suppressed ones, according to [suppressions property](https://docs.oasis-open.org/sarif/sarif/v2.0/csprd02/sarif-v2.0-csprd02.html#_Toc10127852).
+
+### Parse description in comments as `justification`
+
+We propose not to ignore the text after two or more dashes in comments and considered it as the justification of a suppression. The modification should be in `getDirectiveComments`.
+
+### Add suppressions attribute in `LintMessage`
+
+`LintMessage` is the type that finally output to the formatter. Currently it contains `ruleId`, `severity`, `message`, etc. What we desire is a `suppressions` attribute in `LintMessage` when the CLI option `--output-suppression` used. It should be a list that contains a suppression entry (`{kind: "inSource", justification: "Fake justification message."}`).
+
+### Reserve external suppression
+
+In the step of apply external suppression (or disabling rules inside of configuration files), ESLint will not skip rules disabled in configuration files if it gets the `--output-suppression` option. Violations (or problems in ESLint) will be created with `{kind: "external", justification: "Globally disabled"}` if the related rules are switched to `off`.
+
+More specifically, when traversing `configuredRules` in `runRules`, ESLint will not [skip disabled rules](https://github.com/eslint/eslint/blob/81c60f4a8725738f191580646562d1dca7eee933/lib/linter/linter.js#L893) if the `--output-suppression` option used. Then the suppression info `{kind: "external", justification: "Globally disabled"}` will be put into [`createReportTranslator`](https://github.com/eslint/eslint/blob/81c60f4a8725738f191580646562d1dca7eee933/lib/linter/linter.js#L924) and [`createProblem`](https://github.com/eslint/eslint/blob/81c60f4a8725738f191580646562d1dca7eee933/lib/linter/report-translator.js#L357) to create reports with suppressions for disabled rules.
```

Comment:
What should happen if the rule doesn’t exist? Currently, you can have “foo”:”off” in config and it won’t error if the rule “foo” isn’t found. Should we throw an error if a rule isn’t found when in suppression mode?

Reply:
We'd like to keep the current behavior. Thanks.